enhancements/cguConditionEnhancement.md

```diff
@@ -0,0 +1,452 @@
+Point a) and b) can be merged into PreUpgradeValidation where we check managed policies existence, conformity of platform image etc.
+
+ ```yaml
+  Type:    "PreUpgradeValidation",
+  Status:  True/False,
+  Reason:  "MissingManagedPolicy/InvalidPlatformImage/ValidationComplete",
+  Message: “”
+  ```
+
+4. Condition type: **PrecacheSpecValid** and **PrecachingDone** which is already in the current design. For **PrecacheSpecValid**, we may omit the highlighted Reason from the [current spec](#precachingSpec)
+5. Checking the blocking CRs status, if there is any issue UpgradeCannotStart reason is reported. If CRs are not blocked the upgrade can start. Also, adding and deleting labels to the cluster can be performed here. Condition type: Started/Progressing?
+
+ ```yaml
+    Type:    "Started/Progressing",
+    Status:  True/False,
+    Reason:  "MissingBlockingCR/IncompleteBlockingCR/MissingClusterLabel/InProgress",
+    Message: “”
+  ```
+
+  > Mohamed had a comment on this block: "In my opinion those are 2 different conditions 1) Initialized/Started means that the action (copying policies) is happening  2) Progressing/InProgress means that policy remediation started." To discuss.
+
+  6. The **UpgradeCompleted** and **UpgradeTimedOut** Reason can be moved to condition type: Available/Ready
+
+  ```yaml
+    Type:    "Available/Ready",
+    Status:  True/False,
+    Reason:  "UpgradeCompleted/UpgradeTimedOut",
+    Message: “”
+  ``` 
+
+### Workflow summary
+
+- When backup and Precaching is true:
+  * ClusterValidation/ClusterSelected -> BackupDone -> PreUpgradeValidation -> PrecacheSpecValid -> PrecachingDone -> Started/Progressing -> Available/Ready
+ 
+
+- When backup and precaching is false:
+  * ClusterValidation/ClusterSelected -> PreUpgradeValidation -> Started/Progressing -> Available/Ready
+
+### Mapping between Type, Status, Reason and Message:
+
+if backup and/or precaching is false, the related types will not appear in `.status.conditions`.
+
+
+  | Type | Status| Reason| Message |
+  |------|-------|---------|--------|
+  ClusterValidation/ClustersSelected| True| Completed||
+  | |False | NotFound/NotPresent| |
+  `BackupDone` | True | BackupCompleted | Backup is completed |
+  | | False | BackupNotDone | Backup is required and not done |
+  PreUpgradeValidation| True | Completed||
+  | | False | MissingManagedPolicy| |
+  | | False | InvalidPlatformImage| | 
+  `PrecacheSpecValid` | True | PrecacheSpecIsWellFormed | Pre-caching spec is valid and consistent|
+  | | False | NotAllManagedPoliciesExist| The ClusterGroupUpgrade CR has managed policies that are missing|
+  `PrecachingDone` | True | Completed | Precaching is completed |
+  | | False | PrecachingNotDone | Precaching is required and not done | 
+  Started/Progressing| True | InProgress||
+  | | False | MissingBlockingCR | |
+  | | False | IncompleteBlockingCR | | 
+  | | False | MissingClusterLabel | | 
```

Comment:
In what case do we use this reason?

Reply:
In case, we fail to add/delete cluster label, i don't know how important this part is but i can see this functions return error in case labeling goes wrong here: https://github.com/openshift-kni/cluster-group-upgrades-operator/blob/main/controllers/actions.go#L73

Reply:
it should be also false when the below condition Read/Complete in True state

Reply:
You are right.

Reply:
> 

Ah ok, in that case, we don't need a condition. By returning err, we are telling k8s that this is a transient error and try reconcile again in a few seconds. We only need conditions for the stable states where we finish reconcile call normally with a requeue interval.

Reply:
@jc-rh okay